File: core/src/androidTest/java/com/emarsys/core/concurrency/CoreHandlerTest.java
package com.emarsys.core.concurrency;

import android.os.HandlerThread;
import com.emarsys.testUtil.TimeoutUtils;
import org.junit.*;
import org.junit.rules.TestRule;

import java.util.concurrent.CountDownLatch;

public class CoreHandlerTest {

    CoreHandler handler;
    HandlerThread handlerThread;
    Runnable failingRunnable;

    @Rule
    public TestRule timeout = TimeoutUtils.getTimeoutRule();

    @Before
    public void init() {
        String threadName = "test";
        handlerThread = new HandlerThread(threadName);
        handlerThread.start();

        handler = new CoreHandler(handlerThread);

        failingRunnable = new Runnable() {
            @Override
            public void run() {
                throw new RuntimeException("error");
            }
        };
    }

    @After
    public void tearDown() {
        handlerThread.quit();
    }

    @Test
    public void testConstructor_innerLooper_isInitialized() {
        Assert.assertNotNull(handler.getLooper());
        Assert.assertEquals(handlerThread.getName(), handler.getLooper().getThread().getName());
    }

    @Test
    public void testDispatchMessage_shouldBeResilient_toExceptions() throws InterruptedException {
        final CountDownLatch latch = new CountDownLatch(1);

        handler.post(failingRunnable);
        handler.post(failingRunnable);

        handler.post(new Runnable() {
            @Override
            public void run() {
                latch.countDown();
            }
        });

        latch.await();

        Assert.assertTrue(handler.getLooper().getThread().isAlive());
    }

}

File: core/src/androidTest/java/com/emarsys/core/concurrency/CoreHandlerTest.kt
package com.emarsys.core.concurrency

import android.os.HandlerThread
import com.emarsys.testUtil.AnnotationSpec
import io.kotest.matchers.shouldBe
import io.kotest.matchers.shouldNotBe
import java.util.concurrent.CountDownLatch

class CoreHandlerTest : AnnotationSpec() {
    private lateinit var handler: CoreHandler
    private lateinit var handlerThread: HandlerThread
    private lateinit var failingRunnable: Runnable

    @Before
    fun setUp() {
        val threadName = "test"
        handlerThread = HandlerThread(threadName)
        handlerThread.start()
        handler = CoreHandler(handlerThread)
        failingRunnable = Runnable { throw RuntimeException("error") }
    }

    @After
    fun tearDown() {
        handlerThread.quit()
    }

    @Test
    fun testConstructor_innerLooper_isInitialized() {
        handler.looper shouldNotBe null
        handlerThread.name shouldBe handler.looper.thread.name
    }

    @Test
    fun testDispatchMessage_shouldBeResilient_toExceptions() {
        val latch = CountDownLatch(1)
        handler.post(failingRunnable)
        handler.post(failingRunnable)
        handler.post { latch.countDown() }
        latch.await()
        handler.looper.thread.isAlive shouldBe true
    }
}